config/opal_check_package2.m4

```diff
@@ -0,0 +1,232 @@
+AC_DEFUN([OPAL_DECLARE_PACKAGE],[
+    AC_ARG_WITH([$1], [AC_HELP_STRING([--with-$1=DIR], [$2])])
+    AC_ARG_WITH([$1-cppflags], [AC_HELP_STRING([--with-$1-cppflags=FLAGS], [$3])])
+    AC_ARG_WITH([$1-ldflags], [AC_HELP_STRING([--with-$1-ldflags=FLAGS], [$4])])
+    AC_ARG_WITH([$1-libdir], [AC_HELP_STRING([--with-$1-libdir=DIR],
+                                             [Deprecated, use --with-$1-ldflags instead])])
+    AS_IF([test -n "$with_$1_libdir"],
+          [AS_IF([test -n "$with_$1_ldflags"],
+                 [AC_MSG_WARN([It is not possible to use both --with-$1-ldflags and --with-$1-libdir])
+                  AC_MSG_ERROR([Cannot continue])],
+                 [AC_MSG_WARN([--with_$1_libdir=DIR is deprecated, use --with-$1_ldflags=FLAGS instead])
+                  with_$1_ldflags="-L$with_$1_libdir"])])
+])
+
+dnl --------------------------------------------------------------------
+dnl _OPAL_CHECK_PACKAGE2_HEADER(package, prefix, header, includes,
+dnl                             [action-if-found], [action-if-not-found])
+dnl --------------------------------------------------------------------
+AC_DEFUN([_OPAL_CHECK_PACKAGE2_HEADER], [
+    AS_VAR_PUSHDEF([opal_With], [with_$1])
+    AS_VAR_PUSHDEF([opal_Cppflags], [with_$1_cppflags])
+
+    AS_IF([test -n "$opal_Cppflags"],
+          [$2_CPPFLAGS="$$2_CPPFLAGS $opal_Cppflags"
+           CPPFLAGS="$CPPFLAGS $opal_Cppflags"
+           AC_CHECK_HEADERS([$3], [$5], [$6], [$4])],
+          [AS_IF([test -z "$opal_With" || test "$opal_With" = "yes"],
+                 [AC_CHECK_HEADERS([$3], [$5], [$6], [$4])],
+                 [AS_IF([test -r "$opal_With/$3"],
+                        [$2_CPPFLAGS="$$2_CPPFLAGS -I$opal_With"
+                         CPPFLAGS="$CPPFLAGS -I$opal_With"
+                         AC_CHECK_HEADERS([$3], [$5], [$6], [$4])],
+                        [AS_IF([test -r "$opal_With/include/$3"],
+                               [$2_CPPFLAGS="$$2_CPPFLAGS -I$opal_With/include"
+                                CPPFLAGS="$CPPFLAGS -I$opal_With/include"
+                                AC_CHECK_HEADERS([$3], [$5], [$6], [$4])],
+                               [$6])])])])
+
+    AS_VAR_POPDEF([opal_Cppflags])
+    AS_VAR_POPDEF([opal_With])
+])
+
+dnl _OPAL_CHECK_PACKAGE2_LIB(package, prefix, function, library, extra-libraries,
+dnl                          [action-if-found], [action-if-not-found]])
+dnl --------------------------------------------------------------------
+AC_DEFUN([_OPAL_CHECK_PACKAGE2_LIB], [
+    # This is stolen from autoconf to peek under the covers to get the
+    # cache variable for the library check.  one should not copy this
+    # code into other places unless you want much pain and suffering
+    AS_VAR_PUSHDEF([opal_Lib], [ac_cv_search_$3])
+
+    # so this sucks, but there's no way to get through the progression
+    # of search libs without killing off the cache variable and trying
+    # again...
+    unset opal_Lib
```

Comment:
again, we were going to get rid of all this...

Reply:
I am puzzled. do we want to check `DIR/lib64` when there is stuff in `DIR/lib` but `AC_CHECK_LIB` fails ? If yes, I am afraid we do need something like that, though we should update `opal_Lib` (or whatever its new name will be) since we moved from `AC_SEARCH_LIBS` to `AC_CHECK_LIB`. Makes sense ?

Reply:
@ggouaillardet I believe what @bwbarrett is saying is that we agreed not to run ```AC_CHECK_LIB``` at all if they provided a location, either in the ldflags option or in a non-default setting of the "with" option.  Reason is that you'll get a false positive if the lib in the std location passes the test, even if the lib in the specified location fails. So the best you can do is check for existence in lib and lib64, and if found, then set LDFLAGS and move on.

Reply:
Yes, what @rhc54 said.

Reply:
@rhc54 @bwbarrett  let me put it this way.

If an end user `configure --with-FOO=DIR` and in the worst case scenario, you'd rather have configure success (and `make` might either crash because `DIR/lib/libfoo.so` is missing the required symbol, or silently use the libraries in the default location because `DIR/lib/libfoo.so` has the wrong bitness or the linker only accepts static library that is not there) than, silently use a __working__ library in the default location instead or `DIR/lib[64]`.

I fully understand there is not a perfect solution here.

If this is what we want, they I will make the changes, even if I believe this would be a regression (e.g. `make` fails and we could have trapped that in `configure` but we chose not to try).

Reply:
I think you missed the point. Your configure logic is going to succeed anyway because ```AC_CHECK_LIB``` is going to do exactly what you just described - it is a **compile** test, and so the compiler is going to walk thru all those options and find a working library _somewhere_, but not necessarily the one you specified. Setting ```LDFLAGS``` does **not** limit the compiler to looking at just that directory.

Reply:
AC_CHECK_LIB is a link test.  If --with-foo=DIR (or --with-foo-ldflags) found a library, that's the library it found.  That library has the function that's important, in a way that's linkable (has the right size, bitness, static vs. dynamic, etc.).  There's still a versioning problem and if callers are sensitive to a particular version, then they probably need extra tests in the higher level check.

Reply:
@rhc54 I fully understand `AC_CHECK_LIB` __might__ lead to false positives if `libfoo.so` in `DIR/lib` is not usable. (And if it is usable, then this is the one that will be used, not the one in `/usr/lib64`).

13 days ago, you wrote

>  I thought we were going to revise the lib section to avoid use of AC_SEARCH_LIB and use AC_CHECK_LIB instead, and only after checking for existence of the specified lib in the target location?

and yesterday

>  I believe what @bwbarrett is saying is that we agreed not to run AC_CHECK_LIB at all if they provided a location, either in the ldflags option or in a non-default setting of the "with" option.



The only point in my previous comment was if we simply do __not__ run `AC_CHECK_LIB` at all, then we not only keep the same false positives (and there is virtually nothing we can do about that), but we also miss some false negatives (e.g. `configure` successes, but `make` will fail).


And once again, if both of you (@rhc54 and @bwbarrett ) still agree Open MPI should __not__ run `AC_CHECK_LIB` at all (when `--with-FOO-ldflags=LDFLAGS` or `--with-FOO=DIR` is used), then I will implement this change.

Reply:
I have given up - my head has been spun around on this so many times, I've lost count. You guys do whatever you want.